node/src/reactor/main_reactor/control.rs

```diff
@@ -736,135 +722,181 @@ impl MainReactor {
+                            let peers_to_ask = self.small_network.fully_connected_peers_random(
+                                rng,
+                                self.chainspec.core_config.simultaneous_peer_requests as usize,
+                            );
+                            let sync_leap_identifier =
+                                SyncLeapIdentifier::sync_to_historical(parent_hash);
+                            let effects = effect_builder.immediately().event(move |_| {
+                                MainEvent::SyncLeaper(sync_leaper::Event::AttemptLeap {
+                                    sync_leap_identifier,
+                                    peers_to_ask,
+                                })
+                            });
+                            info!("Historical: initiating sync leap for: {}", parent_hash);
+                            KeepUpInstruction::Do(Duration::ZERO, effects)
                         }
+                        LeapStatus::Received {
+                            best_available,
+                            from_peers: _,
+                            ..
+                        } => {
+                            info!(
+                                "Historical: sync leap received for: {:?}",
+                                best_available.trusted_block_header.block_hash()
+                            );
+                            let era_validator_weights = best_available.era_validator_weights(
+                                self.validator_matrix.fault_tolerance_threshold(),
+                            );
+                            for evw in era_validator_weights {
+                                let era_id = evw.era_id();
+                                if self.validator_matrix.register_era_validator_weights(evw) {
+                                    debug!("Historical: got era: {}", era_id);
+                                } else {
+                                    debug!("Historical: already had era: {}", era_id);
+                                }
+                            }
+                            KeepUpInstruction::CheckLater(
+                                "Historical: sync leap received".to_string(),
+                                Duration::ZERO,
+                            )
+                        }
+                        LeapStatus::Awaiting { .. } => KeepUpInstruction::CheckLater(
+                            "Historical: sync leap awaiting".to_string(),
+                            self.control_logic_default_delay.into(),
+                        ),
                     }
-                }
-                BlockSynchronizerProgress::Syncing(_, maybe_block_height, last_progress) => {
-                    error!(
-                        "XXX attempting historical syncing maybe_block_height: {:?}",
-                        maybe_block_height
+                } else if self.block_synchronizer.register_block_by_hash(
```

Comment:
`else`… was this the `self.validator_matrix.has_era(&era_id)` case? :sweat_smile: 
At some point we should extract some of this into smaller methods. (Not necessarily in this PR.)